src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -512,10 +514,59 @@ public static StarlarkInt multiply(StarlarkInt x, StarlarkInt y) {
       return StarlarkInt.of(xl * yl);
     }
 
+    try {
+      long xl = x.toLongFast();
+      long yl = y.toLongFast();
+
+      // see Hacker's Delight, 8-2
+
+      long u0 = xl & 0xFFFFFFFFL;
+      long u1 = xl >> 32;
+      long v0 = yl & 0xFFFFFFFFL;
+      long v1 = yl >> 32;
+      long w0 = u0 * v0;
+      long t = u1 * v0 + (w0 >>> 32);
+      long w1 = t & 0xFFFFFFFFL;
+      long w2 = t >> 32;
+      w1 = u0 * v1 + w1;
+
+      // high and low 64-bits of 128-bit signed multiplication result
+      long high = u1 * v1 + w2 + (w1 >> 32);
+      long low = xl * yl;
+
+      // checking 128-bit signed integer is within 64-bit signed integer range
+      if ((high == 0 && (low & Long.MIN_VALUE) == 0) || (high == -1 && (low & Long.MIN_VALUE) != 0)) {
+        long zl = low | (high & Long.MIN_VALUE);
```

Comment:
I don't think you need this. It sets the sign bit iff high is negative, but we've already established that high consists entirely of copies of low's sign bit.

Reply:
Right. This is leftover from older attempts to implement it correctly. Thanks!